plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java

```diff
@@ -128,6 +133,45 @@ public ILaunch getLaunch(ILaunchConfiguration configuration, String mode) throws
     return super.getLaunch(configuration, mode);
   }
 
+  @Override
+  public boolean finalLaunchCheck(ILaunchConfiguration configuration, String mode,
+      IProgressMonitor monitor) throws CoreException {
+    if (!super.finalLaunchCheck(configuration, mode, monitor)) {
+      return false;
+    }
+
+    // If we're auto-publishing before launch, check if there may be stale
+    // resources not yet published. See
+    // https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1832
+    if (ServerCore.isAutoPublishing()) {
+      IServer server = ServerUtil.getServer(configuration);
+      if (server.shouldPublish() || hasPendingChangesToPublish()) {
+        IStatusHandler prompter = DebugPlugin.getDefault().getStatusHandler(promptStatus);
+        if (prompter != null) {
+          Object continueLaunch = prompter
+              .handleStatus(StaleResourcesStatusHandler.CONTINUE_LAUNCH_REQUEST, configuration);
+          return (Boolean) continueLaunch;
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Check if there are pending changes to be published: this is a nasty condition that can occur if
+   * the user saves changes as part of launching the server.
+   */
+  private boolean hasPendingChangesToPublish() {
```

Comment:
After experimenting, we need to join on the auto build job to ensure the workspace's resource changes happen, so that we see that there are pending events for publishing.